src/program.ts

```diff
@@ -3588,11 +3617,47 @@ export class Function extends TypedElement {
     return local;
   }
 
+  allocateClosedLocal(local: Local) {
+    console.log("marking local " + local.name + " as closed in function " + this.name);
+    local.envOffset = this.nextEnvOffset;
+    // TODO: the alignment logic is slightly more complicated then this iirc
+    this.nextEnvOffset += local.type.byteSize;
+  }
+
+  markClosedLocal(closedLocal: ClosedLocal) {
+    /* Notify the parent that this local must be included in its env. This will also
+     * populate the Local with information about its storage within the Env */
+    if (closedLocal.functionDepth == 1) {
+      this.allocateClosedLocal(closedLocal.local);
+    } else {
+      assert(this.parent.kind == ElementKind.FUNCTION);
+      var parentFunction = <Function>this.parent;
+      parentFunction.markClosedLocal(new ClosedLocal(closedLocal.local, closedLocal.functionDepth - 1));
+    }
+  }
+
   /* @override */
   lookup(name: string): Element | null {
     var locals = this.localsByName;
     if (locals.has(name)) return assert(locals.get(name));
-    return this.parent.lookup(name);
+
+
+    var parentResult = this.parent.lookup(name);
+    if (parentResult !== null) {
+     /* We found a local in a parent of a function, mark that it must be enclosed
+     * We can infer that since this is a Local and not a Closed Local that the function
+     * depth must be 1 */
+      if (parentResult.kind == ElementKind.LOCAL) {
+        assert(this.parent.kind == ElementKind.FUNCTION);
+        /* Return a ClosedLocal with the inferred depth of 1 */
+        return new ClosedLocal(<Local>parentResult)
+      } else if (parentResult.kind == ElementKind.CLOSED_LOCAL) {
+        /* We add 1 to the depth to indicate another level of nesting */
+        var parentClosedLocal = <ClosedLocal>parentResult;
+        return new ClosedLocal(parentClosedLocal.local, parentClosedLocal.functionDepth + 1);
+      }
+    }
+    return parentResult;
```

Comment:
I've moved the state-changing logic out of `lookup`, since as @MaxGraey pointed out it's certainly counterintuitive to have a function called `lookup` changing state. Instead, the caller is expected to call markClosedLocal after receiving a `ClosedLocal` response, as seen above.

Reply:
I like the underlying idea, but I'm also skeptical that `lookup` is the right place for it. For instance, we'd want to deduplicate the same closed local when returned for the second time inside of the same function, both to reduce allocations and to be able to use them as keys in collections, but then we're back to square one because remembering it changes state. As an alternative, an API like `Function#ensureClosedLocal(local: Local): ClosedLocal` that is called after `lookup` upon detecting a local outside of the current function's scope might work. This function would either return the remembered closed local for the given local, or bubble up its parents until it finds the matching local, and only then create a new closed local to return.